### ISSUE TITLE
Restore autoplay on muted video by default or ask for user consent.

### ISSUE THREAD

**Issue by rsepierre** [mozilla-mobile/fenix#16557]
### What is the user problem or growth opportunity you want to see solved?

**Summary** : The current Firefox for android state is by default blocking video autoplay on a "global settings" (for all websites).
It is then most likely blocked for most if not all users.

**Justification :**
The last patch note signaling the an update to the "autoplay" feature dates back to firefox for android version 66.
https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Releases/66
This update states that any media element with audio will be blocked. But this is not the current state of the feature as all media will be blocked whether they have an audio track or not, whether they are muted or volume set to 0.

Every piece of mozilla's documentation states that "muted media element" should normaly autoplay just fine.
Which might signify that the current state or either an oversight or a bug ?
Although it seems presented as a feature hidden in `Menu > Settings > Site Permissions > Automatically read media elements`

**List of problems :**
- The setting is set to "block audio and video" by default
- Mozilla's and MDN and patch notes are not up to date with the current state of the feature. 
- At no time during the installation the user is prompted to opt-in or out of the feature.
- GIF's (1987's technology) are still enabled in firefox, making "storage, performance, data" justification for the block counterproductive (encouraging developpers to use old but still "working" technologies).
- WebGL and canvas (modern technologies) which can be animated by default can be - if used poorly - can be very performance, battery, data or visually demanding. Which makes any "behavioral" or "performance" justifications to block video inconsistent with those features being able to "self start" (autoplay).  
- Web developpers can not ask for user consent to use the feature on their website.
- Users can not set "granular" consent (add a non-invasive website to a whitelist and keep intrusive website at the default setting)
- The setting is very much hidden and users might not even know it is disabled.

### How do you know that this problem exists today? Why is this important?

**The consequences :**
Essentially killing the modern html5 autoplay feature accross the whole internet for any developper who cares about supporting finex OR kills the browser because developers won't support it. This is especially true as all documentation, suggest that firefox perfectly support and enables autoplay (like MDN and caniuse.com)

**Exemple of websites being worse because of it :**
[Giphy.com](https://giphy.com/) (arguably the biggest gif sharing website) is actually using ".wepb" or ".mp4" video instead of GIFs. For firefox they have a fallback using true GIF format. Making the website 10 to 100 times less data efficient on firefox for android than on other browsers.

**Situations where it is preventing websites or usecases from existing :**
Not all video on the web is used as a "film" with video, audio, a start, an end and ment to be watched full lenght.
Just as canvas and webgl the "video" element can have just as much usecase as you can imagine for "things that are animated". While blocking "autoplay" may not be a problem / a good solution for most traditional videos, by blocking the feature by default  (unlike any competitor) and by not providing any way for developpers to ask for consent and users to give website based consent we are essentialy removing support for the feature for all firefox users.
(And if Firefox for android were to go above 0.25% overhall usage developpers might stop implementing it accross the web.)

**More on the subject:**
Google's article on consequences of disabling autoplay and why is it important for developpers to have accurate data on if they can use the feature or not : 
[https://developers.google.com/web/updates/2016/07/autoplay](https://developers.google.com/web/updates/2016/07/autoplay)

Dev.com's article on important benefits of the feature over older (but not blocked by firefox) technologies : 
[https://web.dev/replace-gifs-with-videos/](https://web.dev/replace-gifs-with-videos/)

My pull request to MDN's browser compatibility data ( featured on [caniuse.com - autoplay](https://caniuse.com/?search=autoplay) ) to add a note to notify developpers that firefox for android's users will most likely block every autoplay element :
[https://github.com/mdn/browser-compat-data/pull/7356](https://github.com/mdn/browser-compat-data/pull/7356)

**Tested on :** Firefox for android 82.1.3 (Build #2015774643)

### Any solutions ?

**Feature request:**
Here is multiple ideas / ways we could improve the situation.

- Either rollback and support muted video's autoplay setting by default.
- Add a way for developpers to ask for user consent to autoplay videos on a site to site basis.
- Warns or add visual clue to the users that he is missing the full experience of the interactive and creative website

or alternatively 

- Persist on blocking autoplay on videos to save the world but then to stay true to your principles and also block animated GIF's because it is way worse and maybe also webgl and canvas.

### Who will benefit from it?
Developpers, users, mozilla

**Comment by Amejia481** [comment]
Thanks for filing this ticket @rsepierre, we are working on https://github.com/mozilla-mobile/fenix/issues/11325 to allow users to indicate their autoplay preference per site, and also addressing other related issues https://github.com/mozilla-mobile/fenix/issues/8987

**Comment by rsepierre** [comment]
@Amejia481 Say that site based authorisations and issues are fixed for fenix's to have consistent behavior. Is the intent still to purposefuly block muted (or silent) video from autoplaying **by default** ? And if so, would there be any way the user know he is missing something ? (and thus maybe he would think by himself about whitelisting the website) and/or for the developper to prompt the user for consent ? (Just as you would for using the webcam or other APIs).

**Comment by kbrosnan** [comment]
Closing as a duplicate of #11027

**Comment by Amejia481** [comment]
Let's revise as the issue is more about web compatibility.

**Comment by Amejia481** [comment]
@rsepierre I agree with you, we should change the default setting to just **block audio** content by default as we are doing in Firefox desktop, bellow the default setting on it.

![image](https://user-images.githubusercontent.com/773158/99560662-1d652a80-2994-11eb-8be2-22958a792834.png)


We are working on #11325 to allow users to indicate their autoplay preference per site, also be aware that we have a pending issue for autoplay that depends on this geckoview bug https://bugzilla.mozilla.org/show_bug.cgi?id=1647779

**Comment by Amejia481** [comment]
@kbrosnan we are addressing the change [here](https://github.com/mozilla-mobile/fenix/pull/16640), in case you want to give some feedback :)

**Comment by wisniewskit** [comment]
Just FYI, this is of course causing webcompat reports, for instance https://webcompat.com/issues/61221

**Comment by lobontiumira** [comment]
Verified on the Beta 89.0.0-beta.1 Build, and 4/22 Nightly with Samsung Galaxy Note 8 (Android 9), and Google Pixel (Android 10), on the reparofilms.com page, and on the bristolprintatelier.com, and here are my findings:

* If the user allows audio & video from Settings - Site permissions - Autoplay, both pages start their videos, on mute.

* If the user does not allow audio & video from Settings - Site permissions - Auoplay, the videos don't start, the blue dot is displayed on the padlock icon. The user can select from tapping on it to "Allow audio and video", then the video starts on mute. Then both pages are set as exceptions on Site permissions page in Fenix.
